providers/dns/azure/azure.go

```diff
@@ -200,47 +202,61 @@ func (d *DNSProvider) getHostedZoneID(ctx context.Context, fqdn string) (string,
 	return to.String(zone.Name), nil
 }
 
-// NewServicePrincipalTokenFromCredentials creates a new ServicePrincipalToken using values of the
-// passed credentials map.
-func (d *DNSProvider) newServicePrincipalToken(scope string) (*adal.ServicePrincipalToken, error) {
-	oauthConfig, err := adal.NewOAuthConfig(azure.PublicCloud.ActiveDirectoryEndpoint, d.config.TenantID)
-	if err != nil {
-		return nil, err
-	}
-	return adal.NewServicePrincipalToken(*oauthConfig, d.config.ClientID, d.config.ClientSecret, scope)
-}
-
 // Returns the relative record to the domain
 func toRelativeRecord(domain, zone string) string {
 	return acme.UnFqdn(strings.TrimSuffix(domain, zone))
 }
 
+func getAuthorizer(config *Config) (autorest.Authorizer, error) {
+	if config.ClientID != "" && config.ClientSecret != "" && config.TenantID != "" {
+		oauthConfig, err := adal.NewOAuthConfig(azure.PublicCloud.ActiveDirectoryEndpoint, config.TenantID)
+		if err != nil {
+			return nil, err
+		}
+
+		spt, err := adal.NewServicePrincipalToken(*oauthConfig, config.ClientID, config.ClientSecret, azure.PublicCloud.ResourceManagerEndpoint)
+		if err != nil {
+			return nil, err
+		}
+
+		spt.SetSender(config.HTTPClient)
+		return autorest.NewBearerAuthorizer(spt), nil
+	}
+
+	return auth.NewAuthorizerFromEnvironment()
```

Comment:
somewhere around here is where we would need to set the `HTTPClient` (if it is possible)

Reply:
`NewAuthorizerFromEnvironment` don't use a HTTP client, so it's possible to set the HTTP client.

Reply:
true, but _inside_ that call (https://github.com/Azure/go-autorest/blob/master/autorest/azure/auth/auth.go#L493), you can see they create service principal token (which will make an http call eventually), but they don't seems to expose the token itself. I only bring it up because as this PR stands, it is _slightly_ inconsistent. Meaning if a client assumes that by setting `config.HTTPClient` that all calls would use it, then only in certain paths would that be true.

Reply:
It's possible to set the HTTP client on the Azure client, by example:

```go
	rsc := dns.NewRecordSetsClient(d.config.SubscriptionID)
	rsc.Authorizer = d.authorizer
	rsc.Sender = d.config.HTTPClient
```

And https://github.com/Azure/go-autorest/blob/master/autorest/azure/auth/auth.go#L493 create an URL but don't seems to do a call.

Reply:
Due to this https://github.com/Azure/go-autorest/blob/master/autorest/azure/auth/auth.go#L499, it's possible to set the client on `spToken`

Reply:
I have very little experience with the azure sdk, but can something like this be used? https://github.com/Azure/go-autorest/blob/master/autorest/authorization.go#L149

Reply:
Set the HTTP client with `NewBearerAuthorizer` is already done: L221